### ISSUE TITLE
Test issues

### ISSUE THREAD

**Issue by caiqingnanhai** [XingangPan/SCNN#5]
@XingangPan Hi，
I followed your instruction to make a test and there is no error during test, but the output of the values of all the txt files and the png files are zero, I don't know why and I hope that you can give me some advice, thanks you very much! 
result of txt file like this:
![screenshot from 2018-01-16 10 42 43](https://user-images.githubusercontent.com/20715918/34969521-178be0b6-faaa-11e7-8379-20b62dc9b097.png)

**Comment by XingangPan** [comment]
For some images there is no lane markings, and in this case the values of txt files and png files would be zeros.
You may run the whole test code and see if the F-measure is the same with that in our paper. If it is, there should be no problem. But if **all** txt files and png files are zeros, the F-measure would be zero and something must be wrong. Then you may check if test images are successfully read.

**Comment by caiqingnanhai** [comment]
@XingangPan Hey，
Thanks for your reply and I made a test again, the F-measure shows that the precision is -1, the recall and the Fmeasure are all zero,  I am new to torch, I found that the laneTest.t7 has been generated successfully, I don't know how to check if test images are read successfully, I hope that  you can give me some advices. Thank you!

**Comment by XingangPan** [comment]
You may print some values of the input (like in line 68 of train.py) and see if the values make sense (like they are not constant and lies in [-1.0,1.0]). Then we'll know whether the problem lies in data load part or the network part.
I have not met this problem before, but I'll try to help.

**Comment by caiqingnanhai** [comment]
@XingangPan Thank you so much, the result of the input is like you said that is not constant, but lies in [-10,10], so what should to fix it?
Thank you!

**Comment by XingangPan** [comment]
[-10,10] should be fine. Perhaps take a look at "scoremap" at line 74 of test.py? If it's all zeros, then the problem lies in the model and you may check the output of each layer, otherwise the problem would be in post-processing steps.
You may also set "show" to "True" in line 15 to see if visualization is right. To do so you also need to change "th" to "qtlua" in test.sh.

**Comment by caiqingnanhai** [comment]
@XingangPan Thanks for your reply, I follow your instructions and like you said that the problem maybe in the post-progressing steps, and I use the command `qlua` not `qtlua,` the image visualization is the original image so what should I do next? 
Thank you!

**Comment by XingangPan** [comment]
In the right case the visualization should have probability map of lanes, like:
![visualize](https://user-images.githubusercontent.com/13579537/35101313-12559d98-fc9a-11e7-917d-e9c8f730d41d.png)
So I suspect if the "outputn" in line 89 is all zeros? (In the right case it should have value close to 1.0 at lane marking position) You may print the max value of "outputn" and see whether it's 0 or close to 1.0.

**Comment by caiqingnanhai** [comment]
@XingangPan The value are not zero but very close to zero, the values are all between (0,0.2)

**Comment by XingangPan** [comment]
In my case the max value are close to 1.0 when there are lane markings.
Firstly, you may set -smooth to false at test.sh and see if problem still exists.
Secondly, If problem exists, we may check values of middle layers. To do so, you could:
1. Update the code to the newest version, where we do not shuffle during test.
2. Go to your torch directory, modify torch/install/share/lua/5.1/cudnn/BatchNormalization.lua, add a line at function "updateOutput" to print max value.
![bnprint](https://user-images.githubusercontent.com/13579537/35136932-8d8bf0de-fd21-11e7-8151-8367c3384556.png)
3. Run test code, in my case for the first test image the output is:
![first](https://user-images.githubusercontent.com/13579537/35137139-cda08616-fd22-11e7-95c3-25c258840be9.png)
You may check if our output is the same, then we could better locate the problem.

**Comment by caiqingnanhai** [comment]
@XingangPan when I set `-smooth false` I got the error `data created
data loaded
data loaded
1
time: 2.0890970230103
img: /driver_23_30frame/05151640_0419.MP4/00030.jpg
qlua: bad argument #3 to '?' (torch.*Tensor expected, got userdata)
stack traceback:
	[C]: at 0x7ff9db2bf8d0
	[C]: at 0x7ff9c89f0580
	[C]: in function '__newindex'
	testLane.lua:95: in main chunk
`

**Comment by XingangPan** [comment]
This problem has been fixed. Please clone or pull the newest version.

**Comment by caiqingnanhai** [comment]
@XingangPan The data that I have got is as bellow
![Uploading Screenshot from 2018-01-19 17:26:24.png…]()

**Comment by XingangPan** [comment]
Ok, so the problem exists at or before the first layer.
Could you please print the mean and max of the input?
![printinput](https://user-images.githubusercontent.com/13579537/35144717-56040510-fd40-11e7-8073-f63c947bfe53.png)
In my case it's 0.094 and 2.777 for the first test image.
If yours are the same with me, then even the first convolution layer does not work well, then it seems to be a torch issue and would be a little puzzling... Otherwise would be data loading issue.

**Comment by caiqingnanhai** [comment]
@XingangPan I think I found the problem, there are something wrong with the version of the torch when I make test and I followed the instruction `https://github.com/Fang-Haoshu/multi-human-pose/issues/1`, I modified the file of BatchNormalization.lua, I think this result in that the test errpr, so I want to know that which version that you used for the pretrained model.
thank you very much!

**Comment by XingangPan** [comment]
Hi, sorry to reply late. I installed torch using this repo
https://github.com/torch/distro
And the version is:
[image: 内嵌图片 1]

2018-01-21 13:13 GMT+08:00 caiqingnanhai <notifications@github.com>:
